src/puppetlabs/puppetdb/command.clj

```diff
@@ -86,23 +101,78 @@
    "store report" (version-range 3 6)
    "deactivate node" (version-range 1 3)})
 
-;; ## Command parsing
+(defn- read-command-body
+  [body]
+  (if (= (class body) (class (byte-array 0)))
+    (json/parse-string (String. body "UTF-8") true)
```

Comment:
Should be able to a stick a ByteArrayInputStream on top of body and use `parse-stream` to avoid the intermediate `String`.